### PR TITLE
Initial import - application structure

### PR DESCRIPTION
**Pull request by piontec**
This is a big initial PR that provides:
- basic project setup (still missing: docs, tests)
- basic interfaces for providing build steps
- basic build steps for helm3 backed build pipeline

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,54 @@
+FROM python:3.8.6-alpine3.12 AS base
```
Comment (ubergesundheit) [inline Dockerfile:1]: While usually `alpine` based images are very good, for python it seems it isn't the best choice. See https://pythonspeed.com/articles/alpine-docker-python/ and https://pythonspeed.com/articles/base-image-python-docker-images/

File: Pipfile [hunk 2]
```diff
@@ -12,6 +12,10 @@ bandit = "*"
 flake8 = "*"
 
 [packages]
+configargparse = "*"
+gitpython = "*"
+dependency-injector = "*"
+semver = "*"
```
Comment (ubergesundheit) [inline Pipfile:18]: These should at least pin major versions IMO

File: dabs.sh [hunk 3]
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+
+docker run -it --rm -v `pwd`:/abs/workdir/ app_build_suite:latest $@
```
Comment (ubergesundheit) [inline dabs.sh:3]: Backticks are legacy syntax

```suggestion
docker run -it --rm -v "$(pwd)":/abs/workdir/ app_build_suite:latest $@
```

**Review by ubergesundheit — COMMENTED** [review]
In general looks solid.

In addition to the comments in the code, here are some random thoughts:

- The way `run`, `pre_run` and `cleanup` (sub-)steps are defined and run look rather odd at the first sight. I'm curious on your thoughts about that. Having them defined in their own `BuildStep` descendant seem to make sense, running all `pre_run` then all `run` then all `cleanup` seems odd, you've clearly have something in mind..
- IMO the `Container` class needs another name or some explanation. I immediately thought about a container in the kubernetes/dockersense but that does not seem the case here?

File: Dockerfile (hunk 1 above)
Comment (piontec) [reply Dockerfile:1]: Yeah, I actually had some musl problems before, but never ran into long build time issues. But I think I'll switch to debian-slim.

File: Pipfile (hunk 2 above)
Comment (piontec) [reply Pipfile:18]: Actual versions are pinned in `pipenv.lock`. So far this is highly dev-time version, so I don't think we have to pin versions yet.

File: app_build_suite/build_steps/build_step.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        raise NotImplementedError
+
+    @abstractmethod
+    def cleanup(self, config: argparse.Namespace) -> None:
+        """
+        Clean up any resources used during the BuildStep.
+        :param config: Ready (parsed) configuration Namespace object.
+        :return: None
+        """
+        raise NotImplementedError
```
[the remark this answers is not in the document]
Comment (piontec) [reply app_build_suite/build_steps/build_step.py:84]: You're right, I wanted them to be an interface, but they can still be empty by default. I'll change that.

**Comment by piontec** [comment]
> * The way `run`, `pre_run` and `cleanup` (sub-)steps are defined and run look rather odd at the first sight. I'm curious on your thoughts about that. Having them defined in their own `BuildStep` descendant seem to make sense, running all `pre_run` then all `run` then all `cleanup` seems odd, you've clearly have something in mind..

The problem is shared state: something that is not yet entirely handled in the API. Basically, next step usually uses some form of outcome from a previous step. So, we can't run `cleanup` of the previous step until the full pipeline was executed. As for `pre-run`, it's for "failing fast": if we know we won't be able to execute the build, as something won't work anyway, there's no need to run 90% of actual build to learn that. I'll try to explain it better in the comments.

> * IMO the `Container` class needs another name or some explanation. I immediately thought about a container in the kubernetes/dockersense but that does not seem the case here?

File: app_build_suite/__main__.py [hunk 5]
```diff
@@ -1,8 +1,138 @@
...
+        logging.getLogger().setLevel(logging.DEBUG)
+
+    container = ComponentsContainer()
+    container.config.from_dict({"build_engine": global_only_config.build_engine},)
+
+    steps = get_pipeline(container)
+    config = get_config(steps)
+    run_pre_steps(config, steps)
+    run_build_steps(config, steps)
+    run_cleanup(config, steps)
```
Comment (ubergesundheit) [inline app_build_suite/__main__.py:135]: Would it make sense to always execute `run_cleanup` even if something failed before?

Comment (piontec) [reply app_build_suite/__main__.py:135]: Good idea, probably with a flag informing if the build was OK or fail (you might want to leave something for debugging if the build was failed).

**Review by ubergesundheit — APPROVED** [review]

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    && rm -rf linux-amd64/ && rm helm.tar.gz && helm version
+# ct
+RUN curl -L https://github.com/helm/chart-testing/releases/download/v${CT_VER}/chart-testing_${CT_VER}_linux_amd64.tar.gz \
+    -o ct.tar.gz && tar zxf ct.tar.gz && install ct /usr/local/bin && mkdir /etc/ct \
+    && mv ./etc/* /etc/ct/ && rm ct && rm ct.tar.gz && ct version
+# cleanup
+RUN rm -rf $WORK_DIR
+RUN pip install yamllint==1.25.0 yamale==3.0.4
+
+RUN adduser --disabled-password --home $ABS_DIR --uid 1001 abs
```
Comment (ubergesundheit) [inline Dockerfile:48]: This could be `useradd`

File: app_build_suite/build_steps/build_step.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+                self.name,
+                f"Can't find {bin_name} executable. Please make sure it's installed.",
+            )
+
+    def _assert_version_in_range(
+        self, app_name: str, version: str, min_version: str, max_version: str
+    ) -> None:
+        """
+        Checks if the given app_name with a string version falls in between specified min and max
+        versions (min_version <= version < max_version). Raises ValidationError.
```
Comment (webwurst) [inline app_build_suite/build_steps/build_step.py:121]: Not sure, maybe `max_version` should also be included? Or be renamed?

Comment (piontec) [reply app_build_suite/build_steps/build_step.py:121]: I'll add `_exc`

**Comment by cokiengchiara** [comment]
I assume no need to announce or document for now, moving to done!

**Pull request merged**